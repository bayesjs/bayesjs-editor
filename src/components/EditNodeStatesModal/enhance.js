import {
  branch,
  compose,
  renderNothing,
  withHandlers,
} from 'recompose';
import {
  isNil,
  pipe,
  prop,
} from 'ramda';
import { onCancelEditingNodeStates, onSaveEditingNodeStates } from '@actions/editing-node-states';

import { connectify } from '@decorators';
import { getEditingNodeStates } from '@selectors/editing-node-states';
import { getNodes } from '@selectors';

const hasNotNode = pipe(prop('node'), isNil);

const enhance = compose(
  connectify({
    node: getEditingNodeStates,
    nodes: getNodes,
  }, {
    onSaveNodeStates: onSaveEditingNodeStates,
    onCancel: onCancelEditingNodeStates,
  }),
  branch(hasNotNode, renderNothing),
  withHandlers({
    onSave: ({ onSaveNodeStates, nodes }) => (id, states) => onSaveNodeStates(id, states, nodes),
    onAlert: () => window.alert,
  }),
);

export default enhance;
