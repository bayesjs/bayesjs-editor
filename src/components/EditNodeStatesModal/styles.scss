.edit-states-modal {
  width: 300px;

  &__save-button {
    margin-right: 10px;
  }
}
