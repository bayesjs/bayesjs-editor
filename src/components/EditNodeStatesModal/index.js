import Component from './component';
import enhance from './enhance';

export default enhance(Component);
