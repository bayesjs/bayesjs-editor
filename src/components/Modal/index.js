import PropTypes from 'prop-types';
import React from 'react';
import ReactModal from 'react-modal';
import classNames from 'classnames';
import fontAwesome from 'font-awesome/css/font-awesome.css';
import { getComponentTestId } from '../../utils/test-utils';
import styles from './styles.css';

const style = {
  overlay: {
    backgroundColor: 'rgba(50, 50, 50, 0.75)',
  },
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    transform: 'translate(-50%, -50%)',
    border: 'solid 1px #333',
    borderRadius: 0,
    padding: 0,
    margin: 0,
  },
};

const Modal = ({
  title,
  isOpen,
  onRequestClose,
  children,
  name,
}) => (
  <ReactModal
    isOpen={isOpen}
    onRequestClose={onRequestClose}
    style={style}
    contentLabel={title}
  >
    <div className={styles.header} data-testid={getComponentTestId('ModalHeader', name || title)}>
      {title}
      <i
        onClick={onRequestClose}
        className={classNames([fontAwesome.fa, fontAwesome.faTimes, styles.closeButton])}
        title="Fechar"
      />
    </div>

    <div className={styles.body} data-testid={getComponentTestId('ModalBody', name || title)}>
      {children}
    </div>
  </ReactModal>
);

Modal.defaultProps = {
  children: null,
  name: '',
};

Modal.propTypes = {
  title: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onRequestClose: PropTypes.func.isRequired,
  children: PropTypes.element,
  name: PropTypes.string,
};

export default Modal;
