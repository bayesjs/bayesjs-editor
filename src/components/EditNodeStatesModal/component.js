import React, { useCallback, useState } from 'react';
import {
  append,
  complement,
  includes,
  isEmpty,
  prop,
  remove,
  thunkify,
} from 'ramda';

import Button from '@components/Button';
import Modal from '@components/Modal';
import NodeAddState from '@components/NodeAddState';
import NodeStatesEditList from '@components/NodeStatesEditList';
import PropTypes from 'prop-types';
import { bem } from '@utils/styles';
import { nodePropTypes } from '@models';
import styles from './styles.scss';

const componentClassName = bem(styles);

const propStates = prop('states');
const propId = prop('id');
const notIncludes = complement(includes);
const isNotEmpty = complement(isEmpty);

const onAddStateHandler = (states, setStates, onAlert) => (state) => {
  const isValid = notIncludes(state, states);

  if (isValid) {
    setStates(append(state, states));
  } else {
    onAlert(`O estado "${state}" já foi adicionado.`);
  }

  return isValid;
};

const onSaveHandler = thunkify(({ id }, states, onSave, onAlert) => {
  const isValid = isNotEmpty(states);

  if (isValid) {
    onSave(id, states);
  } else {
    onAlert('Você deve informar pelo menos um estado.');
  }
});

const EditNodeStatesModal = ({
  node,
  onSave,
  onCancel,
  onAlert,
}) => {
  const [states, setStates] = useState(propStates(node));
  const onDeleteState = useCallback(({ index }) => setStates(remove(index, 1, states)), [states]);
  const onAddState = useCallback(
    onAddStateHandler(states, setStates, onAlert),
    [states, setStates, onAlert],
  );
  const onHandleSave = useCallback(
    onSaveHandler(node, states, onSave, onAlert),
    [node, states, onSave, onAlert],
  );

  return (
    <Modal
      title={`Editar Estados (${propId(node)})`}
      onRequestClose={onCancel}
      isOpen
    >
      <div className={componentClassName}>
        <NodeStatesEditList states={states} onDeleteState={onDeleteState} />
        <NodeAddState onAddState={onAddState} />

        <div>
          <Button
            className={componentClassName.element('save-button').toString()}
            onClick={onHandleSave}
            name="save"
            primary
          >
            Salvar
          </Button>
          <Button onClick={onCancel}>Cancelar</Button>
        </div>
      </div>
    </Modal>
  );
};

EditNodeStatesModal.propTypes = {
  node: nodePropTypes,
  onSave: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
  onAlert: PropTypes.func.isRequired,
};

EditNodeStatesModal.defaultProps = {
  node: null,
};

export default EditNodeStatesModal;
