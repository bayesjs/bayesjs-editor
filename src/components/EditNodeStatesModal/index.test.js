import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Modal from 'react-modal';
import EditNodeStatesModal from './component';

const renderComponent = (props) => {
  Modal.setAppElement(document.body);

  return {
    ...render(<EditNodeStatesModal {...props} />),
    element: document.body.lastChild,
  };
};

describe('EditNodeStatesModal Component', () => {
  const defaultProps = {
    node: {
      id: 'node 1',
      states: [
        'True',
        'False',
      ],
      parents: [],
    },
    onSave: jest.fn(),
    onCancel: jest.fn(),
    onAlert: jest.fn(),
  };
  let component;

  beforeEach(() => {
    defaultProps.onSave.mockClear();
    defaultProps.onCancel.mockClear();
    defaultProps.onAlert.mockClear();
    component = renderComponent(defaultProps);
  });

  it('matches snapshot', () => {
    expect(component.element).toMatchSnapshot();
  });

  describe('When add a new state', () => {
    describe('and is a valid one', () => {
      const newState = 'New State';

      beforeEach(() => {
        const { getByPlaceholderText, getByTitle } = component;
        const addStateInput = getByPlaceholderText('Novo estado');
        const addStateButton = getByTitle('Adicionar Estado');

        fireEvent.change(addStateInput, { target: { value: newState } });
        fireEvent.click(addStateButton);
      });

      it('does not call onAlert function', () => {
        expect(defaultProps.onAlert).not.toBeCalled();
      });
    });

    describe('and is a invalid one', () => {
      const newState = 'True';

      beforeEach(() => {
        const { getByPlaceholderText, getByTitle } = component;
        const addStateInput = getByPlaceholderText('Novo estado');
        const addStateButton = getByTitle('Adicionar Estado');

        fireEvent.change(addStateInput, { target: { value: newState } });
        fireEvent.click(addStateButton);
      });

      it('calls onAlert function', () => {
        expect(defaultProps.onAlert).toBeCalledWith(`O estado "${newState}" já foi adicionado.`);
      });
    });
  });

  describe('When clicked in save button', () => {
    describe('and has states', () => {
      beforeEach(() => {
        const { getByText } = component;
        const saveButton = getByText('Salvar');

        fireEvent.click(saveButton);
      });

      it('calls onSave function', () => {
        expect(defaultProps.onSave).toBeCalledWith(['True', 'False']);
      });
    });

    describe('and has no states', () => {
      beforeEach(() => {
        const { getByText, getAllByTitle } = component;
        const saveButton = getByText('Salvar');
        const deleteButtons = getAllByTitle('Remover Estado');

        deleteButtons.forEach(fireEvent.click);
        fireEvent.click(saveButton);
      });

      it('calls onAlert function', () => {
        expect(defaultProps.onAlert).toBeCalledWith('Você deve informar pelo menos um estado.');
      });
    });
  });
});
