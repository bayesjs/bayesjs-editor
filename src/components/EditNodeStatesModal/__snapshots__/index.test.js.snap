// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditNodeStatesModal Component matches snapshot 1`] = `
<div
  class="ReactModalPortal"
>
  <div
    class="ReactModal__Overlay ReactModal__Overlay--after-open"
    style="position: fixed; top: 0px; left: 0px; right: 0px; bottom: 0px; background-color: rgba(50, 50, 50, 0.75);"
  >
    <div
      aria-label="Editar Estados (node 1)"
      class="ReactModal__Content ReactModal__Content--after-open"
      role="dialog"
      style="position: absolute; top: 50%; left: 50%; border: 1px solid #333; background: rgb(255, 255, 255); overflow: auto; border-radius: 0; outline: none; padding: 0px; transform: translate(-50%, -50%); margin: 0px;"
      tabindex="-1"
    >
      <div
        class="header"
        data-testid="modal-header__editar-estados-node-1"
      >
        Editar Estados (node 1)
        <i
          class="fa faTimes closeButton"
          title="Fechar"
        />
      </div>
      <div
        class="body"
        data-testid="modal-body__editar-estados-node-1"
        style="max-width: 819.2px; max-height: 614.4000000000001px;"
      >
        <div
          class=""
        >
          <ul
            class=""
          >
            <li
              class=""
              data-testid="node-state-edit__true"
            >
              <span
                title="True"
              >
                True
              </span>
              <button
                class="__button"
                data-testid="button__remove-state"
                title="Remover Estado"
                type="button"
              >
                <i
                  class="fa fa-trash"
                />
              </button>
            </li>
            <li
              class=""
              data-testid="node-state-edit__false"
            >
              <span
                title="False"
              >
                False
              </span>
              <button
                class="__button"
                data-testid="button__remove-state"
                title="Remover Estado"
                type="button"
              >
                <i
                  class="fa fa-trash"
                />
              </button>
            </li>
          </ul>
          <div
            class=""
          >
            <input
              class="__input"
              placeholder="Novo estado"
              type="text"
              value=""
            />
            <button
              class="__button"
              data-testid="button__adicionar-estado"
              title="Adicionar Estado"
              type="button"
            >
              <i
                class="fa fa-plus"
              />
            </button>
          </div>
          <div>
            <button
              class="__save-button --primary"
              data-testid="button__save"
              title=""
              type="button"
            >
              Salvar
            </button>
            <button
              class=""
              data-testid="button"
              title=""
              type="button"
            >
              Cancelar
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
